src/main/java/com/hivemq/cli/utils/json/JsonMqttPublish.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2019 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.hivemq.cli.utils.json;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSyntaxException;
+import com.hivemq.client.mqtt.datatypes.MqttQos;
+import com.hivemq.client.mqtt.datatypes.MqttTopic;
+import com.hivemq.client.mqtt.datatypes.MqttUtf8String;
+import com.hivemq.client.mqtt.mqtt3.message.publish.Mqtt3Publish;
+import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5PayloadFormatIndicator;
+import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;
+import org.bouncycastle.util.encoders.Base64;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.nio.charset.StandardCharsets;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class JsonMqttPublish extends JsonFormatted {
+
+    @NotNull private final String topic;
+    @NotNull private JsonElement payload;
+    @NotNull private MqttQos qos;
+    @NotNull private String receivedAt;
+    private boolean retain;
+    @Nullable private String contentType;
+    @Nullable private Mqtt5PayloadFormatIndicator payloadFormatIndicator;
+    @Nullable private Long messageExpiryInterval;
+    @Nullable private String responseTopic;
+    @Nullable private String correlationData;
+    @Nullable private Map<String, String> userProperties;
```

Comment:
all fields can be final

Reply:
The Mqtt5 attributes can not be final afaik because they get only defined in one of the constructors?